modules/gapi/include/opencv2/gapi/own/rmat.hpp

```diff
@@ -0,0 +1,52 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2020 Intel Corporation
+
+
+#ifndef OPENCV_GAPI_OWN_RMAT_HPP
+#define OPENCV_GAPI_OWN_RMAT_HPP
+
+#include <opencv2/gapi/gmat.hpp>
+
+namespace cv { namespace gapi { namespace own {
+
+class HostMat
+{
+public:
+    virtual ~HostMat() = default;
+    virtual Mat getMat() const = 0;
+    virtual void flush() const = 0;
+};
+
+class RMat
+{
+public:
+    using HostMatP = std::shared_ptr<HostMat>;
+    enum class AccessType { R, W, RW };
+
+    class Impl
+    {
+    public:
+        virtual ~Impl() = default;
+        virtual GMatDesc desc() const = 0;
+        virtual HostMatP access(AccessType) const = 0;
+    };
+    using ImplP = std::shared_ptr<Impl>;
+
+    RMat() = default;
+    RMat(ImplP impl) : m_impl(impl) {}
+    GMatDesc desc() const { return m_impl->desc(); }
+    HostMatP access(AccessType a) const { return m_impl->access(a); }
```

Comment:
Why it is not a cv::Mat?

Reply:
We need some sort of `flush` method to sync data with remote device but we can't extend cv::Mat with that

Reply:
I'd clearly express that we have a method to obtain data and a method to upload data.
The adapter could also indicate if the obtained data is a copy (then there's no sense in writing to this memory) or a mapping (then this memory is writable). That's it :)

-----

Maybe handling this at user code side is not very useful though. Need to prototype and see how it looks like